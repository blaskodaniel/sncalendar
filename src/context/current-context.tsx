import { ConstantContent, ODataParams } from '@sensenet/client-core'
import { GenericContent } from '@sensenet/default-content-types'
import React, { useEffect, useState } from 'react'
import Semaphore from 'semaphore-async-await'
import { useRepository } from '../hooks/use-repository'
import { useInjector } from '../hooks/use-injector'

export const CurrentContentContext = React.createContext<GenericContent>(ConstantContent.PORTAL_ROOT)
export const CurrentContentProvider: React.FunctionComponent<{
  idOrPath: number | string
  onContentLoaded?: (content: GenericContent) => void
  oDataOptions?: ODataParams<GenericContent>
}> = props => {
  const [loadLock] = useState(new Semaphore(1))
  const [content, setContent] = useState<GenericContent>(ConstantContent.PORTAL_ROOT)
  const repo = useRepository()
  const injector = useInjector()
  const [reloadToken, setReloadToken] = useState(1)
  const reload = () => setReloadToken(Math.random())

  useEffect(() => {
    const events = injector.getEventHub(repo.configuration.repositoryUrl)
    const subscriptions = [
      events.onContentModified.subscribe((c: any) => {
        if (c.content.Id === content.Id) {
          reload()
        }
      }),
    ]
    return () => subscriptions.forEach(s => s.dispose())
  }, [repo, content, injector])

  const [error, setError] = useState<Error | undefined>()

  useEffect(() => {
    const ac = new AbortController()
    if (props.idOrPath) {
      ;(async () => {
        try {
          const response = await repo.load({
            idOrPath: props.idOrPath,
            requestInit: { signal: ac.signal },
            oDataOptions: props.oDataOptions,
          })
          setContent(response.d)
          props.onContentLoaded && props.onContentLoaded(response.d)
        } catch (err) {
          if (!ac.signal.aborted) {
            setError(err)
          }
        } finally {
          loadLock.release()
        }
      })()
    }
    return () => ac.abort()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [repo, props.idOrPath, reloadToken, loadLock])

  if (error) {
    throw error
  }

  return (
    <CurrentContentContext.Provider value={content as GenericContent}>{props.children}</CurrentContentContext.Provider>
  )
}
