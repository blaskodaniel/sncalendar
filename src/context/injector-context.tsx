import { Injector } from '@furystack/inject/dist/Injector'
import { ConsoleLogger } from '@furystack/logging'

import React from 'react'

export const snInjector = new Injector()
snInjector.options.owner = 'SnApp'
snInjector.useLogging(ConsoleLogger)

export const InjectorContext = React.createContext(snInjector)
