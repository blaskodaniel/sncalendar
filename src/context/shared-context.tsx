import React, { createContext, Dispatch, useState } from 'react'
import { GenericContent } from '@sensenet/default-content-types'
import { EditPropertiesDialog } from '../components/edit-dialog'
import CalendarEvent from '../CalendarEvent-type'
import { DialogComponent } from '../components/view-dialog'

export const SharedContext = createContext<{
  openeditmodal: boolean
  setOpeneditmodal: Dispatch<React.SetStateAction<boolean>>
  opendisplaymodal: boolean
  setOpendisplaymodal: Dispatch<React.SetStateAction<boolean>>
  setEvent: Dispatch<React.SetStateAction<GenericContent | undefined>>
}>(null as any)

const SharedProvider: React.FunctionComponent<any> = props => {
  const [openeditmodal, setOpeneditmodal] = useState(false)
  const [opendisplaymodal, setOpendisplaymodal] = useState(false)
  const [event, setEvent] = useState<CalendarEvent>(null as any)

  return (
    <>
      <SharedContext.Provider
        value={{ openeditmodal, setOpeneditmodal, setEvent, opendisplaymodal, setOpendisplaymodal }}>
        {props.children}
        {openeditmodal && event != null ? (
          <EditPropertiesDialog
            content={event}
            dialogProps={{
              open: openeditmodal,
              onClose: () => setOpeneditmodal(false),
              keepMounted: false,
            }}
          />
        ) : null}
        {opendisplaymodal && event != null ? (
          <DialogComponent open={opendisplaymodal} content={event} onClose={() => setOpendisplaymodal(false)} />
        ) : null}
      </SharedContext.Provider>
    </>
  )
}
export default SharedProvider
