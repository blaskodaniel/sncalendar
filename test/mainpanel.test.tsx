import { mount } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { Repository } from '@sensenet/client-core'
import MainPanel from '../src/components/mainpanel'
import { RepositoryContext } from '../src/context/repository-provider'
import { SharedContext } from '../src/context/shared-context'
import { CalendarTestEvent } from './mocks/test-objects'

jest.mock('react-quill')

describe('Mainpanel', () => {
  const content = CalendarTestEvent

  const shareobject = {
    openeditmodal: true,
    setOpeneditmodal: jest.fn(),
    setEvent: jest.fn(),
    opendisplaymodal: false,
    setOpendisplaymodal: jest.fn(),
    event: content,
    refreshcalendar: true,
    setRefreshcalendar: jest.fn(),
    opennewmodal: true,
    setOpennewmodal: jest.fn(),
    setOpennoti: jest.fn(),
  }

  it('Mainpanel snapshot', async () => {
    const repo = new Repository()
    repo.loadCollection = function fetchMethod() {
      return Promise.resolve({ d: { results: [content] } } as any)
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo as any}>
          <SharedContext.Provider value={shareobject as any}>
            <MainPanel />
          </SharedContext.Provider>
        </RepositoryContext.Provider>,
      )
    })
    expect(wrapper).toMatchSnapshot()
  })
})
