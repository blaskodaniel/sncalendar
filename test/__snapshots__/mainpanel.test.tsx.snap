// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mainpanel Mainpanel snapshot 1`] = `
<MainPanel>
  <AddNewEvent>
    <WithStyles(ForwardRef(Fab))
      aria-label="Add new"
      className="makeStyles-fabAdd-6"
      color="secondary"
      onClick={[Function]}
      variant="extended"
    >
      <ForwardRef(Fab)
        aria-label="Add new"
        className="makeStyles-fabAdd-6"
        classes={
          Object {
            "colorInherit": "MuiFab-colorInherit",
            "disabled": "Mui-disabled",
            "extended": "MuiFab-extended",
            "focusVisible": "Mui-focusVisible",
            "label": "MuiFab-label",
            "primary": "MuiFab-primary",
            "root": "MuiFab-root",
            "secondary": "MuiFab-secondary",
            "sizeMedium": "MuiFab-sizeMedium",
            "sizeSmall": "MuiFab-sizeSmall",
          }
        }
        color="secondary"
        onClick={[Function]}
        variant="extended"
      >
        <WithStyles(ForwardRef(ButtonBase))
          aria-label="Add new"
          className="MuiFab-root makeStyles-fabAdd-6 MuiFab-extended MuiFab-secondary"
          component="button"
          disabled={false}
          focusRipple={true}
          focusVisibleClassName="Mui-focusVisible"
          onClick={[Function]}
        >
          <ForwardRef(ButtonBase)
            aria-label="Add new"
            className="MuiFab-root makeStyles-fabAdd-6 MuiFab-extended MuiFab-secondary"
            classes={
              Object {
                "disabled": "Mui-disabled",
                "focusVisible": "Mui-focusVisible",
                "root": "MuiButtonBase-root",
              }
            }
            component="button"
            disabled={false}
            focusRipple={true}
            focusVisibleClassName="Mui-focusVisible"
            onClick={[Function]}
          >
            <button
              aria-label="Add new"
              className="MuiButtonBase-root MuiFab-root makeStyles-fabAdd-6 MuiFab-extended MuiFab-secondary"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiFab-label"
              >
                <ForwardRef
                  className="makeStyles-extendedIcon-7"
                >
                  <WithStyles(ForwardRef(SvgIcon))
                    className="makeStyles-extendedIcon-7"
                  >
                    <ForwardRef(SvgIcon)
                      className="makeStyles-extendedIcon-7"
                      classes={
                        Object {
                          "colorAction": "MuiSvgIcon-colorAction",
                          "colorDisabled": "MuiSvgIcon-colorDisabled",
                          "colorError": "MuiSvgIcon-colorError",
                          "colorPrimary": "MuiSvgIcon-colorPrimary",
                          "colorSecondary": "MuiSvgIcon-colorSecondary",
                          "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                          "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                          "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                          "root": "MuiSvgIcon-root",
                        }
                      }
                    >
                      <svg
                        aria-hidden="true"
                        className="MuiSvgIcon-root makeStyles-extendedIcon-7"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                        />
                      </svg>
                    </ForwardRef(SvgIcon)>
                  </WithStyles(ForwardRef(SvgIcon))>
                </ForwardRef>
                Add new
              </span>
              <NoSsr />
            </button>
          </ForwardRef(ButtonBase)>
        </WithStyles(ForwardRef(ButtonBase))>
      </ForwardRef(Fab)>
    </WithStyles(ForwardRef(Fab))>
  </AddNewEvent>
</MainPanel>
`;
